### PR TITLE
Add config option to automatically detect common multi-line patterns when tailing logs. 

### PR DESCRIPTION
**Pull request by prognant**
### What does this PR do?

NOTE: @gh123man has taken over this feature. Please direct feedback/questions to @gh123man

This PR introduces a new feature for the logs agent to automatically detect common multi line patterns and combine multi-line logs (such as stack traces) into a single line when a pattern is detected. 

The approach used is to scan incoming logs against a list of known common patterns (regexs). If a pattern matches enough messages during the sample window (first 1000 logs or 30 seconds - whichever comes first) the single line handler will be replaced with a multiline handler with the selected pattern. 

When a new log source is detected - the logs are sent to Datadog with the single line handler while the agent tests the known patterns. If a pattern is selected all future logs will be processed with the multi line handler for that source even if the file rotates. However when the agent restarts, the detection will be run again. 

This feature is disabled by default for now under the config `logs_config.auto_multi_line_detection`. 
Additional patterns can be added under the `logs_config.auto_multi_line_extra_patterns` config. 
This feature can also be enabled/disabled per integration with the config `auto_multi_line_detection`. 
There are other options to tweak the behavior of this feature however they should be considered experimental and may change in future versions. 
 
## Setup
1. Create a working directory where your logs are going to be placed. 
2. Set your agent log level to debug
3. Create an integration config for your agent to pick up logs from this directory. Something like:
```yaml
logs:
  - type: file
    path: "/path/to/working/dir/*.log"
    service: "logs-qa"
    source: "logs-qa"
```
4. Create a script called `autorotate.sh` with the following content:
```bash
#!/bin/bash

i=0
while read x; do
    ((i=i+1))
    if ! ((i % $1)); then
        mv autorotate.log autorotate.log.0
    fi
    echo "$x" >> autorotate.log
done
```

## Scenarios 

1. start your agent
2. in your working directory, run `docker run -it bfloerschddog/java-excepton-logger | ./autorotate.sh 1000`
3. tail the agent logs and look for the following log line: 
```
 Pattern ^\d+-\d+-\d+ \d+:\d+:\d+(,\d+)? matched 965 lines with a ratio of 0.965000
```
4. after a few seconds you should see: 
```
 Auto multi line log detection enabled
 Found a previously detected pattern - using multi line handler
 ```
 5. let it run a little longer and this message should be repeated. This means that it is re-using the pattern correctly between file rotations 
 6. look at livetail for your host in app and confirm that the first logs submitted are not multi-line. But the later logs (after the first 1000) should be correctly combined. 

Repeat the above steps with the following containers:
`bfloerschddog/python-exception-logger` - looking for exceptions (like above)
`bfloerschddog/java-every-other-logger` - looking for every log line to be combine (instead of just exceptions)

## Settings and flags

1. test `auto_multi_line_detection` enable/disable globally
2. test `auto_multi_line_detection` enable/disable per integration 
3. test `auto_multi_line_extra_patterns`
 - add a new regex to the list in global config `\d+\.\d+\.\d+\.\d+` (to match IP addresses)
 - feed the agent logs that match this regex: `docker run -it --rm mingrammer/flog -l | ./autorotate.sh 1000`
    (note flog's default logs start with an IP address)
 - confirm the regex is used by looking for a log line (confirm it is indeed your new custom pattern): 
 ```
 Pattern ^\d+\.\d+\.\d+\.\d+ matched 1000 lines with a ratio of 1.00000
  ```
  Note: flog only produces single line logs so no need to look for multi-line combine logs. just testing that the user regex worked. 

 4. set `container_collect_all: true` and test that logs read from the docker socket also match a pattern. 
  - run `docker run -it bfloerschddog/java-every-other-logger`
  - look at agent logs for: `Pattern ^[A-Za-z_]+ \d+, \d+ \d+:\d+:\d+ (AM|PM) matched 500 lines with a ratio of 0.500000`

### CODE REVIEW

File: pkg/logs/decoder/line_handler.go [hunk 1]
```diff
@@ -246,3 +253,210 @@ func (h *MultiLineHandler) sendBuffer() {
...
+}
+
+// AutoMultilineHandler can switch from single to multiline handler if upon the occurrence
+// of a stable pattern at the begginning of the process
+type AutoMultilineHandler struct {
+	multiLineHandler  *MultiLineHandler
+	singleLineHandler *SingleLineHandler
+	inputChan         chan *Message
+	outputChan        chan *Message
+	flipChan          chan struct{}
```
Comment (djmitche) [inline pkg/logs/decoder/line_handler.go:289]: what does this property do?

File: releasenotes/notes/multiline-auto-sensing-daa189b46e547de3.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    It checks for the 1000 first lines (or waits 30 seconds)
+    when tailing for a know list of timestamp format. If the 
+    number of matched lines is greater than the threshold it 
+    switches to the MultiLineHandler with the regexp matching
+    the timestamp format. The regex chosen is saved in the log
+    config and is reused if the file rotates.  New global config 
+    parameter: ``logs_config.auto_multi_line_detection`` to enable
+    the feature.  New per log integration config parameters:
+    * ``auto_multi_line_detection`` to enable the feature on a case
+      by case basis
```
Comment (djmitche) [inline releasenotes/notes/multiline-auto-sensing-daa189b46e547de3.yaml:21]:
```suggestion
    Add multi-line auto-sensing when tailing logs from file.
    It checks the 1000 first lines (or waits 30 seconds, whichever is first)
    when tailing for a list of known timestamp formats. If the 
    number of matched lines is greater than the threshold it 
    switches to the MultiLineHandler with the pattern matching
    the timestamp format. The pattern chosen is saved in the log
    config and is reused if the file rotates.  Use the new global config 
    parameter: ``logs_config.auto_multi_line_detection`` to enable
    the feature.  New per log integration config parameters:
    * ``auto_multi_line_detection`` to enable the feature on a case
      by case basis
```

**Review by djmitche — APPROVED** [review]

File: pkg/logs/decoder/line_handler.go (hunk 1 above)
Comment (gh123man) [reply pkg/logs/decoder/line_handler.go]: It's used to stop the auto multiline handlers main loop and halt it's parsing when it switches over to a mutli line handler. I can add a comment or rename it if it's unclear.

File: pkg/logs/decoder/line_handler.go [hunk 3]
```diff
@@ -246,3 +253,210 @@ func (h *MultiLineHandler) sendBuffer() {
...
+}
+
+// Get gets the pattern
+func (d *DetectedPattern) Get() *regexp.Regexp {
+	d.Lock()
+	defer d.Unlock()
+	return d.pattern
+}
+
+// AutoMultilineHandler can switch from single to multiline handler if upon the occurrence
```
[the remark this answers is not in the document]
Comment (gh123man) [reply pkg/logs/decoder/line_handler.go]: Im going to reword this.

File: pkg/logs/decoder/line_handler.go (hunk 1 above)
Comment (djmitche) [reply pkg/logs/decoder/line_handler.go]: Got it.  A comment would be useful.  It looks like this is only used within the `run` goroutine, where `run` -> `processAndTry` -> `switchToMultilineHandler`.  Could this be replaced with a boolean that's consulted in the loop in `run`?

Comment (gh123man) [reply pkg/logs/decoder/line_handler.go]: Yep good point. I changed it to `isRunning` which is easier to understand.

File: pkg/logs/decoder/auto_multiline_handler.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+func (h *AutoMultilineHandler) processAndTry(message *Message) {
+	// Process message before anything else
+	h.singleLineHandler.process(message)
+
+	for i, scoredPattern := range h.scoredMatches {
+		match := scoredPattern.regexp.Match(message.Content)
+		if match {
+			log.Tracef("A regexp matched during multiline auto sensing: %v", scoredPattern.regexp)
```
Comment (remeh) [inline pkg/logs/decoder/auto_multiline_handler.go:141]: It's `Trace` so we are fine for regular usage but, wouldn't this be flooding the output log if a file is filled with a pattern? I'm just wondering if we should not log a trace (or debug) when leaving this for if `scoredPattern.score` has changed?
I may be overcautious here, please ignore if I'm exaggerating.

File: pkg/logs/decoder/auto_multiline_handler.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+		log.Debug("Multiline auto detect timed out before reaching line test threshold")
+		timeout = true
+		break
+	default:
+		break
+	}
+
+	h.linesTested++
+	if h.linesTested >= h.linesToAssess || timeout {
+		topMatch := h.scoredMatches[0]
```
Comment (remeh) [inline pkg/logs/decoder/auto_multiline_handler.go:167]: Quick check: there is no branch where `scoredMatches` could be empty?

File: pkg/logs/decoder/auto_multiline_handler.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+func (h *AutoMultilineHandler) switchToMultilineHandler(r *regexp.Regexp) {
+	h.isRunning = false
+	h.singleLineHandler = nil
+
+	// Build and start a multiline-handler
+	h.multiLineHandler = newMultiLineHandler(h.inputChan, h.outputChan, r, h.flushTimeout, h.lineLimit)
+	h.multiLineHandler.Start()
+
+	// At this point control is handed over to the multiline handler and the AutoMultilineHandler read loop has stopped.
```
Comment (remeh) [inline pkg/logs/decoder/auto_multiline_handler.go:190]: Perfect comment, thanks. 👌

File: pkg/logs/decoder/decoder.go [hunk 7]
```diff
@@ -121,11 +145,44 @@ func NewDecoderWithEndLineMatcher(source *config.LogSource, parser parser.Parser
...
+
+func buildAutoMultilineHandlerFromConfig(outputChan chan *Message, lineLimit int, source *config.LogSource, detectedPattern *DetectedPattern) *AutoMultilineHandler {
+	linesToSample := source.Config.AutoMultiLineSampleSize
+	if linesToSample <= 0 {
+		linesToSample = dd_conf.Datadog.GetInt("logs_config.auto_multi_line_default_sample_size")
+	}
+	matchRatio := source.Config.AutoMultiLineMatchThreshold
+	if matchRatio == 0 {
+		matchRatio = dd_conf.Datadog.GetFloat64("logs_config.auto_multi_line_default_match_threshold")
+	}
```
Comment (remeh) [inline pkg/logs/decoder/decoder.go:159]: Nit: should we rename it `matchThreshold`?

**Review by remeh — APPROVED** [review]
Great PR! Made a few comments but LGTM as-is already.

File: pkg/logs/decoder/auto_multiline_handler.go (hunk 5 above)
Comment (gh123man) [reply pkg/logs/decoder/auto_multiline_handler.go:167]: 👍 Nope. The hardcoded list of patterns is always set to the `scoredMatches` prop, and they are never evicted from the list so this is safe.

File: pkg/logs/decoder/auto_multiline_handler.go (hunk 4 above)
Comment (gh123man) [reply pkg/logs/decoder/auto_multiline_handler.go:141]: you are right. ill remove this log line - there are other useful log lines later on for the cumulative match results.

**Pull request merged**